### ISSUE TITLE
Clearer documentation

### ISSUE THREAD

**Issue by pbhogale** [greta-dev/greta#144]
I was not sure that these points had been covered in earlier issues, so opening a new one. 
As a data scientist new to PPLs, I found the documentation hard to follow. 

In particular :

1. Where does the "inference" happen exactly ? What is the connection between 
`distribution(y) <- normal(mean, sd)` which expresses the relationship between the parameters (whose priors have been defined) and the target variable (the data we want to explain/fit to) and 
`m <- model(int, coef, sd)` which is only a function of the parameters. How does `m` know about `y` or the structure of the model ? (it clearly does, since `plot(m)` plots the structure of the model).. so clearly, `distribution(y) <- normal(mean, sd)` and `m <- model(int, coef, sd)` are intimately connected... but how ?

2. It would be excellent to have examples in the form of full (even if small) workflows. i.e.
- experimental data with a clearly understood relationship between predictors and target variable  
- plausible mathematical model for target variable in terms of predictors
- greta implementation of said model  
- greta fit to the experimental data  
- some measure of the "goodness of fit".. at least by inspection, where one might plot the model with the inferred parameters against raw data. 

3. Some explanation of where in the workflow tensorflow is used and how this is beneficial. An example with a non-trivial data set (larger than would be practicable without tensorflow) would make the benefits of the tf backend explicit. 

4. A function to `predict` would be a good idea. Given a model that has been fitted (again, do the learned/fitted parameters live in the model structure ? do they live in `draws` ? can the learned parameters be put into the model to generate data ?) the user should be able to save the model (as a .h5 file, ideally) and generate the target variable from this model, given values of the predictor variables. 

I love the project (I came to it via the Tensorflow talk JJ Allaire gave at rstudio::conf2018) and would love to see it blossom, congratulations on some excellent work. 

I worked out a simple linear regression in Greta just to get myself to understand Greta better. https://theclarkeorbit.github.io/greta-playground.html.

**Comment by goldingn** [comment]
Hi Prasanna, thank you for the feedback!

I'm very keen to improve the documentation and help people understand what's going on. So getting input like this is invaluable. I have some specific plans which I think will address most of these points, which I've listed below. Any comments on those - *from anyone reading this* - would be much appreciated.

1. *understanding the black magic*

I'll try and answer this question in as few words as possible, to see if I can write a clearer description:

> greta arrays *know* which other greta arrays they have been created from, and which ones they are used to create; an invisible web of connections between them is created as you define your model. Once you have finished writing out the model code, the greta arrays you care about need to be bundled up into an object that can be used for inference. When you call `model()`, you specify the greta arrays you care about and of which you want to obtain samples or estimates, and the invisible connections between greta arrays are used to find all of the other relevant greta arrays - that's why `y` is spotted and appears in the plot. The model object contains those greta arrays, the connections between them (in the form of a [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph)), and some other ancillary information. The model object then has all the information it needs to construct the corresponding tensorflow graph, and provides (internally) functions to send candidate parameters to the tensorflow graph, and return calculated quantities from the graph. These functions are what are needed to do inference like MCMC or optimisation. The model then acts as the interface between the model-agnostic inference methods (`mcmc()` and `opt()`) and the web of greta arrays, so you can do inference on your model.

(see the P.S. below)

I have tried to provide a complete breakdown of how to *use* these components, without understanding the internal details [here](https://greta-dev.github.io/greta/get_started.html). And I've provided more technical information on the internal workings [here](https://greta-dev.github.io/greta/technical_details.html). Both of these are also distributed as package vignettes.

It seems like an intermediate explanation of this black magic (like I've attempted above) would be useful either in the *get started* vignette, or elsewhere. I'd be really keen to know if there were any particular parts of those vignettes that you found hard to interpret, or whether there were particular things that are missing from them.

2. *examples*

There is a [section on the website](https://greta-dev.github.io/greta/example_models.html) (also distributed as a vignette in the package) for example models. This is pretty empty so far, and obviously needs to be filled in. The wonderful @softloud is planning to code up some more examples to go in that section. The examples there so far are focussed on helping people translate code/approaches from BUGS and Stan, though putting up simple examples for common types of model (e.g. hierarchical components, latent variables etc.) would probably be more helpful (there's an issue for this #44).

I love the idea of making whole worked analyses available! Your example is brilliant, and I can see that that would be way more accessible for understanding how to set up the whole model than just the core code. I think those whole workflows should go in a separate **section on the website for case studies**.

3. *what tensorflow is doing*

The short answer is that all of the underlying computation to calculate the model density and its gradient is done by tensorflow, so that will be most beneficial where there are lots of big matrix/vector operations. That's a bit intangible to most users though, so:
I plan on providing **a section on the website with benchmarks** of greta (on a laptop/CPU server/GPU) versus other available packages on a set of models of varying sizes. That should help potential users work out whether it's worth switching to greta, or something else they already know.

I think it would be beneficial to have some **documentation on how to write efficient greta code**; ie. avoiding for loops and using vectorisation so that tensorflow can do it's thing. That will probably be a separate vignette.

4. *a predict function*

This is done, and it just landed on the Github master branch! It's called `calculate()` (to avoid name conflicts with existing methods, and because that covers cases where its use isn't for *prediction* persay). You can calculate posterior samples for an arbitrary greta array given the output of mcmc, even if that greta array was defined after the model was constructed. The helpfile is [here](https://greta-dev.github.io/greta/calculate.html) any suggestions to help users understand it would be much appreciated.

To make sure these things stay on my radar, I've created issues for:
**a case studies website section** #145
**a benchmarking website section** #146
and **a vignette/website section on writing efficient greta code** #147

Getting the documentation right is a priority for me, and after a several months of distractions I now have some time to focus more on greta. However, there is a lot of work to be done, including lots of stuff in the code internals that would be hard for anyone else to tackle. So I would appreciate any help from the community in improving this documentation.

If anyone would like to take on one of these issues; fix unclear text in the existing documentation; or provide some example models, please let me know and we can coordinate efforts!

---

P.S. In an early prototype of greta, there was no `model` step, you could just create  greta arrays and then pass them straight to the sampler. But that was very counterintuitive to the early testers, who wanted a single tangible object representing the model. The model object also enables things like plotting that would otherwise by difficult.

**Comment by pbhogale** [comment]
Thank you for that detailed and _very_ helpful response. It certainly clarified a lot of things.

I have a couple of suggestions -

1.

The getting started section is too complicated. I would suggest this section only have 
- installation   
- importing data (with an example)   
- building a model and plotting the computational graph (the DAG) with the explicit statement that Greta builds this graph internally when underlying data structures (starting with the model parameter priors) are modified all the way to the output tensor. Link to a seperate article on Greta arrays, converting to and from, etc  
- bayesian inference using `mcmc` (with a mention that the gradients needed by the hamiltonian MC are computed by tensorflow making large models tractable in Greta)  
- max likelihood (or other) optimisation using `opt`
- bayesian prediction with `calculate`
This will mean that the getting started page will help people actually do a small, useful thing with Greta while getting an overview of what greta is doing, with links for further exploration.

 2.

The technical details article should introduce/motivate the basics a little bit, with a paragraph on tensors and DAGs representing operations on tensors. Then, it would be nice to have an explanation of why greta arrays are built as higher level objects over tensors and nodes, and what properties they have, etc. In general each section of this document seems very loosely connected to the one before. Motivating how section (n+1) is connected to what came before in section (n) would help beginners like me follow along. I would also think that this section should emphasize concepts rather than actually display source code, which anyone interested can look up on github. This might also be the right place to explain the role and advantages of the tensorflow back end, since this is the huge advantage Greta has over rstan, winbugs etc. 

3. 

The examples in the examples section should be relatively self contained and motivated with some text. While I understand that comparisons to stan and winbugs are important since those communities have potential greta users, perhaps there can be a seperate page explicitly comparing greta models to stan models and performance metrics, differences in expressiveness etc. the examples page of the greta could be showcase of useful use-cases and code a la https://shiny.rstudio.com/gallery/ or https://matplotlib.org/examples/index.html.

I would - of course - be happy to assist in any way I can. I am completely new to PPls, but I hope to get up to speed quickly and implement a couple of non-trivial examples soon.

**Comment by goldingn** [comment]
That's really helpful, thank you! I'll digest and try to block out some time to blitz the documentation in the not too distant future
